### PR TITLE
feat: enable ec_add / ec_mul

### LINKED ISSUE #1230: ecAdd

### LINKED ISSUE #1231: ecMul

### LINKED ISSUE #1365: dev: remove with_attr ecadd / ecmul
Remove those checks 
* https://github.com/kkrt-labs/kakarot/blob/b3f8e1864c410fb35d689cdcad8fc271694e4c86/src/kakarot/precompiles/ecadd.cairo#L46
* https://github.com/kkrt-labs/kakarot/blob/b3f8e1864c410fb35d689cdcad8fc271694e4c86/src/kakarot/precompiles/ecmul.cairo#L45

No errors are raised in `alt_bn128_g1.cairo`

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1230
Resolves #1231 
Resolves #1365

## What is the new behavior?

- Adds ECMUL and ECADD execution from Cairo1Helpers
- Fixes an issue where __reverted__ precompile execution was not consuming all gas due to a mismanagement of error codes
- note: reverted != execution fails silently (see ec_recover vs. ec_add, for example)
- Fixes an issue where external_precompiles would not fail with error EXCEPTIONAL_HALT

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1398)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,8 +6,8 @@ endif
 
 .PHONY: build test coverage
 
-# 154615699 corresponds to release v0.1.7 of Kakarot SSJ.
-KKRT_SSJ_RELEASE_ID = 154615699
+# 173874367 corresponds to release v0.1.12 of Kakarot SSJ.
+KKRT_SSJ_RELEASE_ID = 173874367
```
Comment (ClementWalter) [inline Makefile:10]: maybe time to do https://github.com/kkrt-labs/kakarot/issues/1394

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol
Comment (ClementWalter) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:1]: we had already some precompiles in `PlainOpcodes.sol`, can me migrate them here then?

See
```
    function kill(address payable to) public payable {
        selfdestruct(to);
    }

    function mulmodMax() public pure returns (uint256) {
        return mulmod(type(uint256).max, type(uint256).max, type(uint256).max);
    }

    function addmodMax() public pure returns (uint256) {
        return addmod(type(uint256).max, type(uint256).max, type(uint256).max);
    }

    function computeHash(bytes memory input) public returns (bytes32) {
        bytes32 hash = keccak256(input);
        emit HashComputed(msg.sender, hash);
        return hash;
    }
```

File: pyproject.toml [hunk 2]
```diff
@@ -8,7 +8,7 @@ packages = [{ include = "kakarot_scripts" }, { include = "tests" }]
 
 [tool.poetry.dependencies]
 python = ">=3.10,<3.11"
-starknet-py = "^0.23.0"
+starknet-py = { git = "https://github.com/kkrt-labs/starknet.py", rev = "cc272228b66c84079b01034f0d205aae8e7bc113" }
```
Comment (ClementWalter) [inline pyproject.toml:11]: 🥵 can't we stick to a regular version?

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        async def test_should_return_ec_mul_with_coordinates(self, evm_precompiles):
+            expected_x = (
+                0x30644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFD3
+            )
+            expected_y = (
+                0x15ED738C0E0A7C92E7845F96B2AE9C0A68A6A449E3538FC7FF3EBF7A5A18A2C4
+            )
+            assert await evm_precompiles.ecMul(1, 2, 2) == (expected_x, expected_y)
+
+        async def test_should_return_ec_mul_point_at_infinity(self, evm_precompiles):
```
Comment (ClementWalter) [inline tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:42]: use diff testing with reference implementation from https://github.com/ethereum/execution-specs/blob/master/src/ethereum/shanghai/vm/precompiled_contracts/alt_bn128.py

```
from ethereum.shanghai.vm.precompiled_contracts.alt_bn128 import ...
```

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol
Comment (enitrat) [reply solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:1]: these aren't precompiles? 🤔

File: pyproject.toml (hunk 2 above)
Comment (ClementWalter) [reply pyproject.toml:11]: leave 0.23 until we face an issue

**Comment by ClementWalter** [comment]
paused until katana update their blockifier fork

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /// @dev Gas cost for ECADD call is 150
+    uint256 private constant ECADD_GAS = 1000000;
+    /// @dev Gas cost for ECMUL call is 6000
+    uint256 private constant ECMUL_GAS = 1000000;
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+    constructor() {}
```
Comment (obatirou) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:21]: No need to explicitly write the default constructor. You can omit it as it is empty

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    /// @notice Performs elliptic curve addition
+    /// @param x1 X coordinate of the first point
+    /// @param y1 Y coordinate of the first point
+    /// @param x2 X coordinate of the second point
+    /// @param y2 Y coordinate of the second point
+    /// @return success True if the operation was successful, false otherwise
+    /// @return x X coordinate of the result point
+    /// @return y Y coordinate of the result point
+    function ecAdd(uint256 x1, uint256 y1, uint256 x2, uint256 y2)
+        public
```
Comment (obatirou) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:35]: change visibility to external for function that are called only from external contract and not internally

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    /// @notice Performs elliptic curve scalar multiplication
+    /// @param x1 X coordinate of the point
+    /// @param y1 Y coordinate of the point
+    /// @param s Scalar for multiplication
+    /// @return success True if the operation was successful, false otherwise
+    /// @return x X coordinate of the result point
+    /// @return y Y coordinate of the result point
+    function ecMul(uint256 x1, uint256 y1, uint256 s) public view returns (bool success, uint256 x, uint256 y) {
```
Comment (obatirou) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:55]: change visibility to external for function that are called only from external contract and not internally

File: kakarot_scripts/utils/starknet.py [hunk 7]
```diff
@@ -243,6 +243,12 @@ def get_deployments():
 
 @cache
 def get_artifact(contract_name):
+    # Cairo 0 artifacts
```
Comment (enitrat) [inline kakarot_scripts/utils/starknet.py:246]: Get CairoZero before Cairo 1 to avoid eventual conflicts

**Comment by enitrat** [comment]
image
final commit this time i swear

File: Makefile [hunk 8]
```diff
@@ -6,11 +6,11 @@ endif
 
 .PHONY: build test coverage clean
 
-# 154615699 corresponds to release v0.1.7 of Kakarot SSJ.
-KKRT_SSJ_RELEASE_ID = 154615699
+# 173874367 corresponds to release v0.1.13 of Kakarot SSJ.
+KKRT_SSJ_RELEASE_ID = 176384150
```
Comment (ClementWalter) [inline Makefile:10]: not the same number

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+contract EvmPrecompiles {
+    /// @dev Address of the ECADD precompile
+    address private constant ECADD_PRECOMPILE = address(0x06);
+    /// @dev Address of the ECMUL precompile
+    address private constant ECMUL_PRECOMPILE = address(0x07);
+
+    /// @dev Gas cost for ECADD call is 150
+    uint256 private constant ECADD_GAS = 1000000;
+    /// @dev Gas cost for ECMUL call is 6000
+    uint256 private constant ECMUL_GAS = 1000000;
```
Comment (ClementWalter) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:16]: I don't understand the comment wrt the values

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    //////////////////////////////////////////////////////////////*/
+    /// @notice Performs elliptic curve addition
+    /// @param x1 X coordinate of the first point
+    /// @param y1 Y coordinate of the first point
+    /// @param x2 X coordinate of the second point
+    /// @param y2 Y coordinate of the second point
+    /// @return success True if the operation was successful, false otherwise
+    /// @return x X coordinate of the result point
+    /// @return y Y coordinate of the result point
+    function ecAdd(uint256 x1, uint256 y1, uint256 x2, uint256 y2) external view returns (bool, uint256 x, uint256 y) {
```
Comment (ClementWalter) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:29]: why adding a `bool` in the output ?

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        bytes memory input = abi.encodePacked(x1, y1, x2, y2);
+        (bool success, bytes memory result) = ECADD_PRECOMPILE.staticcall{gas: ECADD_GAS}(input);
+        if (!success) {
```
Comment (ClementWalter) [inline solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:32]: what is result when success is false? empty?

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+
+from kakarot_scripts.utils.kakarot import deploy
+
+
+@pytest_asyncio.fixture(scope="package")
+async def evm_precompiles(owner):
+    return await deploy(
+        "EvmPrecompiles",
+        "EvmPrecompiles",
+        caller_eoa=owner.starknet_contract,
```
Comment (ClementWalter) [inline tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:17]: if the `msg.sender` is not used in the constructor, you can remove this
```suggestion
```

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    )
+
+
+def ref_alt_bn128_add(x0, y0, x1, y1):
```
Comment (ClementWalter) [inline tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:21]: add docstring for the source of this code

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+
+    for i in (x0_value, y0_value, x1_value, y1_value):
+        if i >= ALT_BN128_PRIME:
+            raise OutOfGasError
+
+    try:
+        p0 = BNP(BNF(x0_value), BNF(y0_value))
+        p1 = BNP(BNF(x1_value), BNF(y1_value))
+    except ValueError:
+        raise OutOfGasError from None
```
Comment (ClementWalter) [inline tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:35]: let's just remove the try/except then and keep the valueerror

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+    x = Uint(int.from_bytes(x_bytes, "big"))
+    y = Uint(int.from_bytes(y_bytes, "big"))
+
+    return [x, y]
+
+
+@pytest.mark.asyncio(scope="package")
+@pytest.mark.EvmPrecompiles
+# @pytest.mark.xfail(reason="Katana doesn't support new builtins")
```
Comment (ClementWalter) [inline tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:75]:
```suggestion
```

File: solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol (hunk 11 above)
Comment (enitrat) [reply solidity_contracts/src/EvmPrecompiles/EvmPrecompiles.sol:32]: yes

File: tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py (hunk 14 above)
Comment (ClementWalter) [reply tests/end_to_end/EvmPrecompiles/test_evm_precompiles.py:35]: may be easier to read the test if the ref implementation has the same signature, meaning it returns (bool x, y) and don't raise

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**